### PR TITLE
[CPU] EmbeddingBagOffsetsSum, EmbeddingBagPackedSum, EmbeddingSegmentsSum operations.

### PR DESCRIPTION
**Pull request by nshchego**

### CODE REVIEW

File: ngraph/src/ngraph/opsets/opset0_tbl.hpp [hunk 1]
```diff
@@ -98,7 +98,10 @@ NGRAPH_OP(DynPad, ngraph::op)
 NGRAPH_OP(DynReplaceSlice, ngraph::op)
 NGRAPH_OP(DynSlice, ngraph::op)
 NGRAPH_OP(Elu, ngraph::op)
+NGRAPH_OP(EmbeddingBagOffsetsSum, ngraph::op)
```
Comment (lazarevevgeny) [inline ngraph/src/ngraph/opsets/opset0_tbl.hpp:101]: Please, revert the changes. These operations are not a part of opset0.

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+    if (_with_weights)
+        weights_data = inputs[PER_SAMPLE_WEIGHTS_IDX]->cbuffer().as<const T*>();
+    init_from_inputs(inputs);
+
+    const auto& inDataDims = inputs[0]->getTensorDesc().getDims();
+    const size_t IN_DATA_DIMS_SIZE = inDataDims.size();
+
+    const size_t OUTPUT_BAGS_NUM = outputs[0]->getTensorDesc().getDims()[0];
+
+    memset(dst_data, 0, outputs[0]->byteSize());
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp:269]: Can we do initialization "inplace" with main "emb_cycle"? Otherwise memset assumes an extra memory pass (through all dst tensor).

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/embedding_bag_offsets_sum.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+#include "common_test_utils/test_constants.hpp"
+
+using namespace LayerTestsDefinitions;
+
+namespace {
+
+const std::vector<InferenceEngine::Precision> netPrecisions = {
+        InferenceEngine::Precision::FP32,
+//        InferenceEngine::Precision::U8,
+//        InferenceEngine::Precision::I8
```
Comment (dmitry-gorokhov) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/embedding_bag_offsets_sum.cpp:17]: I think we should test other precisions as well

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+        if (inData == nullptr || indicesData == nullptr)
+            THROW_IE_EXCEPTION << log_prefix << _l_name << "' layer has nullable input data.";
+
+        const auto precision = inData->getTensorDesc().getPrecision();
+        if (!supported_precisions.empty()) {
+            if (supported_precisions.find(precision) == supported_precisions.end())
+                THROW_IE_EXCEPTION << log_prefix << _l_name << "' layer has unsupported precision: " << precision.name();
+        } else {
+            static const std::set<Precision> default_supported_precisions =
+                {Precision::FP32, Precision::I8, Precision::U8, Precision::I32};
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp:160]: We need to force FP32 in case input or output have BF16 precision.

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+                 THROW_IE_EXCEPTION << log_prefix << _l_name << "' layer must have equal shapes for indices and per_sample_weights inputs.";
+        }
+
+        LayerConfig config;
+        config.inConfs.resize(layer->insData.size());
+        for (int i = 0; i < layer->insData.size(); i++) {
+            auto data = layer->insData[i].lock();
+            if (data == nullptr)
+                THROW_IE_EXCEPTION << log_prefix << _l_name << "' layer has nullable input data";
+            config.inConfs[i].desc = TensorDesc(data->getTensorDesc());
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp:181]: You should directly specify planar layout since implementation supports only it.

**Comment by dmitry-gorokhov** [comment]
@lazarevevgeny Are you OK with reference impls inside Ngraph?

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp (hunk 2 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp:269]: Due to there is element wise sum operation below, all elements should be initialized.

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+    memset(dst_data, 0, outputs[0]->byteSize());
+
+    std::function<void(size_t, size_t, size_t, size_t, bool)> emb_cycle =
+        [&](size_t src_index, size_t dst_index, size_t emb_idx, size_t weights_idx, bool with_weights) {
+            for (size_t i = 0lu; i < inDataDims[emb_idx]; i++) {
+                size_t new_src_idx = src_index + i * _multipliers[emb_idx];
+                size_t new_dst_idx = dst_index + i * _multipliers[emb_idx];
+                if (emb_idx == IN_DATA_DIMS_SIZE - 1) {
+                    if (with_weights)
+                        dst_data[new_dst_idx] += src_data[new_src_idx] * weights_data[weights_idx];
```
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp:278]: Weight value depends on index value, i.e. a1 * x + a2 * y

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp (hunk 4 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp]: Fixed

File: ngraph/src/ngraph/runtime/reference/embedding_bag_offsets_sum.hpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
+// Copyright (C) 2018-2020 Intel Corporation
```
[the remark this answers is not in the document]
Comment (nshchego) [reply ngraph/src/ngraph/runtime/reference/embedding_bag_offsets_sum.hpp]: Fixed

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+                            dst_data[dst_index + i] += src_data[src_index + i];
+                    }
+                    if (with_weights)
+                        weights_idx++;
+                }
+            }
+        }
+    };
+
+    if (emb_bag_kernel) {
```
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp:365]: As you can see from performance counters: \\nnt-srv01.inn.intel.com\shareD\Users\nshchego\models\common\dlrm\performanc, jit implementation may gain layer up to 6X.

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/embedding_bag_offsets_sum.cpp (hunk 3 above)
Comment (nshchego) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/embedding_bag_offsets_sum.cpp]: Fixed

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    T* dst_data = outputs[0]->buffer().as<T*>() +
+        outputs[0]->getTensorDesc().getBlockingDesc().getOffsetPadding();
+    const T* weights_data = nullptr;
+    if (_with_weights)
+        weights_data = inputs[PER_SAMPLE_WEIGHTS_IDX]->cbuffer().as<const T*>();
+    init_from_inputs(inputs);
+
+    const auto& inDataDims = inputs[0]->getTensorDesc().getDims();
+
+    const size_t OUTPUT_BAGS_NUM = outputs[0]->getTensorDesc().getDims()[0];
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp:140]: Why caps?

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+            auto data = layer->insData[i].lock();
+            if (data == nullptr)
+                THROW_IE_EXCEPTION << log_prefix << _l_name << "' layer has nullable input data";
+            config.inConfs[i].desc = TensorDesc(data->getTensorDesc());
+        }
+
+        DataConfig outConfig;
+        outConfig.desc = TensorDesc(inData->getTensorDesc().getPrecision(),
+            layer->outData[0]->getTensorDesc().getDims(),
+            layer->outData[0]->getTensorDesc().getLayout());
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp:74]: Output layout have to be equal to the input one.

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_offset_sum.cpp [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+                        + std::string(inputs[1]->getTensorDesc().getPrecision().name()) + "'";
+                errorMsg.copy(resp->msg, sizeof(resp->msg) - 1);
+            }
+                return GENERAL_ERROR;
+            }
+        }
+    }
+
+    template<typename T, typename I>
+    StatusCode process_data(
```
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/embedding_bag_offset_sum.cpp:92]: Here is no init_from_inputs() function, which casts integer inputs to size_t. It is virtual and cannot be with template parameters. In this solution function process_data parameterized by 2 type arguments.  
Excluding of this function gave some performance improvement.

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp (hunk 5 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp]: Fixed

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    }
+}
+
+StatusCode MKLDNNEmbeddingBagSum::execute(
+        std::vector<Blob::Ptr>& inputs,
+        std::vector<Blob::Ptr>& outputs,
+        ResponseDesc *resp) noexcept {
+    switch (inputs[0]->getTensorDesc().getPrecision()) {
+        case Precision::FP32:
+        case Precision::BF16: {
```
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp]: Fixed

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp (hunk 9 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp]: Modified

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp]: Removed

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+            for (size_t i = 0lu; i < inDataDims[emb_idx]; i++) {
+                size_t new_src_idx = src_index + i * _multipliers[emb_idx];
+                size_t new_dst_idx = dst_index + i * _multipliers[emb_idx];
+                if (emb_idx == IN_DATA_DIMS_SIZE - 1) {
+                    if (with_weights)
+                        dst_data[new_dst_idx] += src_data[new_src_idx] * weights_data[weights_idx];
+                    else
+                        dst_data[new_dst_idx] += src_data[new_src_idx];
+                } else {
+                    emb_cycle(new_src_idx, new_dst_idx, emb_idx + 1, weights_idx, with_weights);
```
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp]: Removed

File: inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp (hunk 10 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/embedding_bag_sum.cpp]: FIxed

**Comment by dmitry-gorokhov** [comment]
@nshchego  I need green CI to merge this.

**Comment by rkazants** [comment]
I am running W&D model against these changes today. Let me do this before merge please.

**Comment by nshchego** [comment]
> I am running W&D model against these changes today. Let me do this before merge please.

@rkazants , have you checked this model?

**Comment by rkazants** [comment]
> > I am running W&D model against these changes today. Let me do this before merge please.
> 
> @rkazants , have you checked this model?

I got this error in run-time: "data [linear/linear_model/linear_model/linear_model/age_bucketized/weighted_sum/SplitForIndices3919_const] doesn't exist
C:\Users\rkazants\openvino\inference-engine\src\legacy_api\src\cnn_network_impl.cpp:302
C:\Users\rkazants\openvino\inference-engine\include\details/ie_exception_conversion.hpp:62"

**Comment by rkazants** [comment]
> > > I am running W&D model against these changes today. Let me do this before merge please.
> > 
> > 
> > @rkazants , have you checked this model?
> 
> I got this error in run-time: "data [linear/linear_model/linear_model/linear_model/age_bucketized/weighted_sum/SplitForIndices3919_const] doesn't exist
> C:\Users\rkazants\openvino\inference-engine\src\legacy_api\src\cnn_network_impl.cpp:302
> C:\Users\rkazants\openvino\inference-engine\include\details/ie_exception_conversion.hpp:62"

After some hacks in IR I managed to test W&D model and results match with reference. Now the bug is on MO side for correct IR generation.
